src/Razor/src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -21,11 +22,19 @@ internal partial class HtmlMarkupParser : TokenizerBackedParser<HtmlTokenizer>
             SyntaxFactory.Token(SyntaxKind.OpenAngle, "<"),
         };
 
+        private Stack<TagTracker> _tagTracker = new Stack<TagTracker>();
+
         public HtmlMarkupParser(ParserContext context)
             : base(context.ParseLeadingDirectives ? FirstDirectiveHtmlLanguageCharacteristics.Instance : HtmlLanguageCharacteristics.Instance, context)
         {
         }
 
+        private TagTracker CurrentTracker => _tagTracker.Count > 0 ? _tagTracker.Peek() : null;
+
+        private string CurrentStartTagName => CurrentTracker?.TagName;
+
+        private SourceLocation CurrentStartTagLocation => CurrentTracker?.TagLocation ?? SourceLocation.Undefined;
```

Comment:
Unused here but I need this to show errors in ParseBlock case.

Reply:
nit: Consider grouping all of this into a private class

Reply:
There already is a private class called `TagTracker`. These are just convenience properties